src/query/subquery_cache.c

```diff
@@ -0,0 +1,444 @@
+  k1 = (SQ_KEY *) key1;
+  k2 = (SQ_KEY *) key2;
+  sz1 = k1->n_elements;
+  sz2 = k2->n_elements;
+  assert (sz1 == sz2);
+
+  for (i = 0; i < sz1; i++)
+    {
+      if (!mht_compare_dbvalues_are_equal (k1->dbv_array[i], k2->dbv_array[i]))
+	{
+	  return 0;
+	}
+    }
+  return 1;
+
+}
+
+/*
+ * sq_rem_func () - Function to remove an entry from the scalar subquery cache.
+ *   return: NO_ERROR on success.
+ *   key(in): The key of the entry to remove.
+ *   data(in): The data associated with the key.
+ *   args(in): Additional arguments (unused).
+ *
+ * This function is called when an entry is removed from the scalar subquery cache. It frees the resources
+ * allocated for the key and the data (SQ_VAL structure) using sq_free_key and sq_free_val functions.
+ */
+int
+sq_rem_func (const void *key, void *data, void *args)
+{
+  sq_free_key ((SQ_KEY *) key);
+  sq_free_val ((SQ_VAL *) data);
+  return NO_ERROR;
+}
+
+/*
+ * sq_cache_initialize () - Initializes the cache for a given XASL node.
+ *   return: NO_ERROR if successful, ER_FAILED otherwise.
+ *   xasl(in/out): The XASL node for which the cache is being initialized.
+ *
+ * This function creates a hash table for caching the results of the XASL node. It sets up initial values for cache hit and miss
+ * counters and marks the cache as initialized. The function returns NO_ERROR upon successful initialization, or ER_FAILED if the
+ * hash table could not be created.
+ */
+int
+sq_cache_initialize (THREAD_ENTRY * thread_p, XASL_NODE * xasl)
+{
+  UINT64 max_subquery_cache_size = (UINT64) prm_get_bigint_value (PRM_ID_MAX_SUBQUERY_CACHE_SIZE);
+  int sq_hm_entries = (int) max_subquery_cache_size / 2048;	// default 1024
+
+  xasl->sq_cache->ht = mht_create ("sq_cache", sq_hm_entries, sq_hash_func, sq_cmp_func);
+  if (!xasl->sq_cache->ht)
+    {
+      return ER_FAILED;
+    }
+  xasl->sq_cache->enabled = true;
+  xasl->sq_cache->size_max = max_subquery_cache_size;
+  return NO_ERROR;
+}
+
+/*
+ * sq_put () - Puts a value into the cache for a given XASL node.
+ *   return: NO_ERROR if the value is successfully cached, ER_FAILED otherwise.
+ *   xasl(in): The XASL node for which the value is being cached.
+ *   regu_var(in): The regu variable containing the value to be cached.
+ *
+ * This function attempts to cache the result of a regu variable associated with a given XASL node. It generates a key based on
+ * the XASL node's structure and creates a cache entry if such a key does not already exist in the cache. The function returns
+ * NO_ERROR if the value is successfully cached, or ER_FAILED if the key could not be generated or the value could not be added
+ * to the cache.
+ */
+int
+sq_put (THREAD_ENTRY * thread_p, SQ_KEY * key, XASL_NODE * xasl, REGU_VARIABLE * regu_var)
+{
+  SQ_VAL *val;
+  const void *ret;
+  UINT64 new_entry_size = 0;
+  int i;
+
+  val = sq_make_val (thread_p, regu_var);
+
+  if (!xasl->sq_cache->ht)
+    {
+      sq_cache_initialize (thread_p, xasl);
```

Comment:
Error handling required. And it seems better to change the location as shown below.
```
qexec_start_mainblock_iterations  <== sq_cache_initialize 
qexec_end_mainblock_iterations  <== sq_cache_destroy
```

Reply:
If d`estroy` is added to `end_mainblock_iteration`, the issue of statistical information not being aggregated occurs. I will consider adding `init` to `start_mainblock_iteration`.